apps/trading/pages/markets/index.page.tsx

```diff
@@ -1,77 +1,63 @@
-import assign from 'assign-deep';
-import { useApolloClient } from '@apollo/client';
+import merge from 'lodash/merge';
 import { useRouter } from 'next/router';
 import { AsyncRenderer } from '../../components/async-renderer';
 import { MarketListTable, getRowNodeId } from '@vegaprotocol/market-list';
 import {
   Markets_markets,
   Markets_markets_data,
-  MarketsDataProviderCallbackArg,
   marketsDataProvider,
 } from '@vegaprotocol/graphql';
+import { useDataProvider } from '@vegaprotocol/react-helpers';
 
 import type { AgGridReact } from 'ag-grid-react';
 
 const Markets = () => {
   const { pathname, push } = useRouter();
-  const [markets, setMarkets] = useState<Markets_markets[]>(undefined);
-  const [loading, setLoading] = useState<boolean>(true);
-  const [error, setError] = useState<Error>(undefined);
-  const client = useApolloClient();
   const gridRef = useRef<AgGridReact>();
-  const initialized = useRef<boolean>(false);
-
-  useEffect(() => {
-    return marketsDataProvider(
-      client,
-      ({ data, error, loading, delta }: MarketsDataProviderCallbackArg) => {
-        setError(error);
-        setLoading(loading);
-        if (!error && !loading) {
-          if (!initialized.current || !gridRef.current) {
-            initialized.current = true;
-            setMarkets(data);
-          } else {
-            const update: Markets_markets[] = [];
-            const add: Markets_markets[] = [];
-
-            // split into updates and adds
-            if (!gridRef.current) return;
-            const rowNode = gridRef.current.api.getRowNode(
-              getRowNodeId(delta.market)
-            );
-
-            if (rowNode) {
-              const updatedData = produce(
-                rowNode.data.data,
-                (draft: Markets_markets_data) => assign(draft, delta)
-              );
-              if (updatedData !== rowNode.data.data) {
-                update.push({ ...rowNode.data, data: delta });
-              }
-            }
-            // @TODO - else add new market
-            if (update.length || add.length) {
-              gridRef.current.api.applyTransactionAsync({
-                update,
-                add,
-                addIndex: 0,
-              });
-            }
-          }
+  const update = useCallback(
+    (delta: Markets_markets_data) => {
+      const update: Markets_markets[] = [];
+      const add: Markets_markets[] = [];
+      if (!gridRef.current) {
+        return false;
+      }
+      const rowNode = gridRef.current.api.getRowNode(
+        getRowNodeId(delta.market)
+      );
+      if (rowNode) {
+        const updatedData = produce<Markets_markets_data>(
+          rowNode.data.data,
+          (draft: Markets_markets_data) => merge(draft, delta)
+        );
+        if (updatedData !== rowNode.data.data) {
+          update.push({ ...rowNode.data, data: updatedData });
         }
       }
-    );
-  }, [client, initialized]);
+      // @TODO - else add new market
```

Comment:
todo?

Reply:
We need to handle corner case where new market is added. This is not critical because list of markets will be not changing very often right now, but it need to be fixed.